sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala

```diff
@@ -284,8 +292,12 @@ class MempoolMap(
       }
     }
 
-    nonces.put(account, newExecTxs.lastKey.add(BigInteger.ONE))
-    executableTxs.put(account, newExecTxs)
+    if (newExecTxs.nonEmpty) {
+      nonces.put(account, newExecTxs.lastKey.add(BigInteger.ONE))
+      executableTxs.put(account, newExecTxs)
+    } else
+      nonces.put(account, txsFromRejectedBlocks.head.getNonce)
```

Comment:
Yes, correct. As you said, now it may be possible that even the first re-applied transaction is rejected, because it may have an invalid (for our node) size => all the remaining transactions (if any) will be non executable and the next expected nonce will be equal to the nonce of the first re-applied transaction